src/main.rs

```diff
@@ -85,33 +58,80 @@ async fn run_single_slot(
+    rx: mpsc::Receiver<()>,
+) -> Result<()> {
+    ReceiverStream::new(rx)
+        .for_each_concurrent(MAX_CONCURRENT_SLOTS, |_| async {
+            if let Some(slot) = slot_provider.get_last_slot().await {
+                log::debug!(
+                    "processing slot with number: {}, slot start: {}, processing started at: {}",
+                    slot.number,
+                    slot.to_date_time(),
+                    Utc::now()
+                );
+                timeout(
+                    tokio::time::Duration::from_secs(RUN_SLOT_LIMIT_SECS),
+                    run_single_slot(
+                        price_provider.clone(),
+                        message_generator.clone(),
+                        message_broadcaster.clone(),
+                        slot,
+                    ),
+                )
+                .await
+                .unwrap_or_else(|_| {
+                    log::error!(
+                        "Hit {}s timeout for slot: {}",
+                        RUN_SLOT_LIMIT_SECS,
+                        slot.number
+                    );
+                    Ok(())
+                })
+                .unwrap_or_else(|e| {
+                    log::error!("Error when running for slot: {} - {:?}", slot.number, e);
+                });
+            }
+        })
+        .await;
+    Ok(())
+}
+
 #[tokio::main]
 async fn main() -> Result<()> {
     env_logger::init();
 
-    let mut gofer_cmd = std::env::var("GOFER_CMD")?; // Assumes that GOFER_CMD env variable is set
-                                                     // to the gofer binary
+    // Assumes that GOFER_CMD env variable is set to the gofer binary
+    let mut gofer_cmd = std::env::var("GOFER_CMD")?;
     gofer_cmd.push_str(" prices --norpc ETH/USD");
     log::debug!("Gofer command: {}", gofer_cmd);
 
-    let price_provider = GoferPriceProvider::new(&gofer_cmd);
+    let price_provider = Arc::new(GoferPriceProvider::new(&gofer_cmd));
     log::info!("Initialized price_provider");
     // TODO: Replace with a signature provider that lets the operator use their validator key
     let signature_provider = PrivateKeySignatureProvider::random();
     log::info!("Initialized signature_provider");
     let message_generator = MessageGenerator::new(Box::new(signature_provider));
     log::info!("Initialized message_generator");
-    let http_broadcaster = HttpMessageBroadcaster::new(None)?;
+    let message_broadcaster = Arc::new(HttpMessageBroadcaster::new()?);
     log::info!("Initialized message_roadcaster");
-    // TODO: Replace with a provider that returns every slot number independent of whether it's been mined
-    let slot_provider = SystemClockSlotProvider::new();
+    // It's hard to have the slot provider be sent across threads, own half of the notification
+    // channel, and be able to drop the other half of the channel when the slot provider is
+    // dropped. Room for improvement here.
+    let (tx, rx) = mpsc::channel(1);
```

Comment:
Not too happy with how the channel shows up all the way over here in main. I'm sure we can do better.

Reply:
It is rather low level, but nothing that bothers me too much tbh.